paddle/framework/type_inference.md

```diff
@@ -0,0 +1,101 @@
+# Type (or Shape) Inference
+
+## Starting from Examples
+
+```python
+x = layer.data("images", input_size=64*64)
+y = layer.fc(x, output_size=100, activation=softmax)
+l = layer.data("label", dims=10)
+cost = layer.cross_entropy(y, l)
+```
+
+In this example, given the known size of input as 64*64, we can infer shapes of other variables:
+
+1. x is a tensor of shape `[UNK, 64*64]`, where UNK is the minibatch size.
+1. W created by layer.fc is a tensor of shape `[64*64, 100]`
+1. b created by layer.fc is a tensor `[100]`
+1. x*W is an intermediate tensor of shape `[UNK, 100]`
+1. x*W+b is an intermediate tensor of shape `[UNK, 100]`
+1. y = softmax(x*W+b) is a tensor of shape `[UNK, 100]`
+1. l is a tensor of shape `[UNK, 10]`
+1. cost is a tensor of shape `[UNK, 1]`
+
+It is possible that we have variable-length inputs, e.g., text of words:
+
+```python
+x = layer.data("paragraph", lod_level=2, input_size=6000)
+```
+
+`lod_level` specifies that a paragraph is composed of sentences, which consists of words.  `input_size` is the dictionary size -- we have 6000 known words.  In this case,
+
+- x is an LODTensor of shape `[UNK, UNK, 6000], lod_level=2`.
+
+If the input is a video, or a sequence of images:
+
+```python
+x = layer.data("video", lod_level=1, input_size=640*480)
+```
+
+- x is an LODTensor of shape `[UNK, UNK, 640*480], lod_level=1`.
+
+## VarDesc
+
+```protobuf
+message VarDesc {
+  enum Type {
+    INT = 0;
+    FLOAT = 1;
+    STRING = 2;
+    INTS = 3;
+    FLOATS = 4;
+    STRINGS = 5;
+    LOD_TENSOR = 6;
+  }
+  Type type = 1;
+  optional LODTesnorDesc lod_tensor = 2; // when type==LOD_TENSOR
```

Comment:
LOD_TENSOR just a description of tensor's sequence information. It could be float lod tensor,int lod tensor, etc.